internal/handlers/take_backup_retry.go

```diff
@@ -0,0 +1,220 @@
+		return nil
+	}
+
+	switch r := config.Retries.(type) {
+	case *pb.RetryConfig_Count:
+		{
+			if int(r.Count) == ops {
+				return nil
+			}
+			res := lastEnd.Add(exp(ops) * MIN_BACKOFF)
+			return &res
+		}
+	case *pb.RetryConfig_MaxBackoff:
+		{
+			bck := exp(ops) * MIN_BACKOFF
+			if bck >= r.MaxBackoff.AsDuration() {
+				return nil
+			}
+			res := lastEnd.Add(bck)
+			return &res
+		}
+	default:
+		return nil
+	}
+}
+
+func HandleTbOps(config *pb.RetryConfig, tbOps []*types.TakeBackupOperation) (int, error) {
+	//select last tbOp.
+	//if nothing, run new, skip
+	//if there is a tbOp, check its status
+	//if success: set success to itself
+	//if cancelled: set error to itself
+	//if error:
+	//if we can retry it: retry, skip
+	//if no more retries: set error to itself
+	if len(tbOps) == 0 {
+		return RunNewTb, nil
+	}
+	last := tbOps[len(tbOps)-1]
+	switch last.State {
+	case types.OperationStateDone:
+		return Success, nil
+	case types.OperationStateError:
+		{
+			t := shouldRetry(config, len(tbOps), last.Audit.CompletedAt.AsTime())
+			if t != nil && time.Now().After(*t) {
+				return RunNewTb, nil
+			}
+			return Error, nil
+		}
+	case types.OperationStateRunning:
+		return Skip, nil
+	}
+	return Error, errors.New("unexpected tb op status")
+}
+
+func TBWROperationHandler(
+	ctx context.Context,
+	operation types.Operation,
+	db db.DBConnector,
+	clientConn client.ClientConnector,
+	s3 config.S3Config,
+	clientConfig config.ClientConnectionConfig,
+) error {
+	xlog.Info(ctx, "TBWROperationHandler", zap.String("OperationID", operation.GetID()))
+
+	if operation.GetType() != types.OperationTypeTBWR {
+		return fmt.Errorf("wrong operation type %s != %s", operation.GetType(), types.OperationTypeTBWR)
+	}
+	tbwr, ok := operation.(*types.TakeBackupWithRetryOperation)
+	if !ok {
+		return fmt.Errorf("can't cast Operation to TakeBackupWithRetryOperation %s", types.OperationToString(operation))
+	}
+
+	ops, err := db.SelectOperations(ctx, queries.NewReadTableQuery(
+		queries.WithTableName("Operations"),
+		queries.WithIndex("idx_pc"),
+		queries.WithQueryFilters(queries.QueryFilter{
+			Field:  "parent_operation_id",
+			Values: []table_types.Value{table_types.StringValueFromString(tbwr.ID)},
+		}),
+		queries.WithOrderBy(queries.OrderSpec{
+			Field: "created_at",
+		}),
+	))
+	tbOps := make([]*types.TakeBackupOperation, len(ops))
+	for i := range ops {
+		tbOps[i] = ops[i].(*types.TakeBackupOperation)
+	}
+	if err != nil {
+		return fmt.Errorf("can't select Operations for TBWR op %s", tbwr.ID)
+	}
+
+	switch tbwr.State {
+	case types.OperationStateRunning:
+		{
+			do, err := HandleTbOps(tbwr.RetryConfig, tbOps)
+			if err != nil {
+				tbwr.State = types.OperationStateError
+				tbwr.Message = err.Error()
+				errup := db.ExecuteUpsert(ctx, queries.NewWriteTableQuery().WithUpdateOperation(tbwr))
+				if errup != nil {
+					return errup
+				}
+				//I am kinda lost here. When should we return err and when should we xlog.Error()?
```

Comment:
help pls

Reply:
you should return error here, it should be logged by processor